indexer/indexer.go

```diff
@@ -0,0 +1,247 @@
+// Package indexer contains functions for sourcing and aggregating article counts. It makes heavy use of the
+// github.com/zavitax/sortedset-go implementation for ranking.  It depends on a Storage implementation for caching
+// and a fetcher function for retrieving the articles from the original source
+// TODO: these api calls have a mostly similar algorithm that could be refactored into a higher-order function
+package indexer
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"github.com/zavitax/sortedset-go"
+	"io"
+	"net/http"
+	"pelotechfun/constants"
+	"pelotechfun/messages"
+	"pelotechfun/storage"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// Type fetcher is an internal type that describes a standard function for fetching day counts from an external source
+type fetcher = func(date time.Time) ([]messages.ArticleCount, error)
+
+var (
+	//Var Fetcher holds an instance of a fetcher function. It is exported to enable  stubbing for tests
+	Fetcher fetcher = wikipediafetcher
+	//Var DB is a cache for article day counts.  It is exported to enable stubbing for tests
+	DB storage.Storage = storage.NewLocalMapStorage()
+)
+
+// wikipediafetcher is a wrapper fetcher function for the Wikipedia Pageviews API.
+func wikipediafetcher(date time.Time) ([]messages.ArticleCount, error) {
+	counts := []messages.ArticleCount{}
+	year := strconv.Itoa(date.Year())
+	month := date.Format(constants.TWODAYMONTH)
+	day := date.Format(constants.TWODAYDAYOFWEEK)
+	url := fmt.Sprintf(constants.PAGEVIEWS_URL, year, month, day)
```

Comment:
Fetcher should be moved to its own package as the feature is distinct from the indexing functions below.

From a design point of view we should consider moving wikipedia fetching and caching functionality to its own service.  This will allow us to scale indexing with less impact on wikipedia (i.e. running up against rate limiting) as well as being able to share cached records among indexing instances.